syftlib/src/main/java/org/openmined/syft/Syft.kt

```diff
@@ -1,85 +1,139 @@
-                        )
-                        is NetworkMessage.SocketError -> Log.e(TAG, "socket error", it.throwable)
-                        is NetworkMessage.MessageReceived -> handleResponse(
-                            CommunicationDataFactory.deserializeSocket(
-                                it.message
-                            )
-                        )
-                        is NetworkMessage.MessageSent -> println("Message sent successfully")
+        return job
+    }
+
+    fun requestCycle(job: SyftJob) {
+        if (this::workerId.isInitialized)
+            socketClient.getCycle(
+                CycleRequest(
+                    workerId,
+                    job.modelName,
+                    job.version,
+                    getPing(),
+                    getDownloadSpeed(),
+                    getUploadSpeed()
+                )
+            ).compose(schedulers.applySingleSchedulers())
+        else {
+            compositeDisposable.add(socketClient.authenticate()
+                    .compose(schedulers.applySingleSchedulers())
+                    .subscribe { t: AuthenticationSuccess ->
+                        if (!this::workerId.isInitialized)
+                            setSyftWorkerId(workerId)
+                        requestCycle(job)
                     }
-                }
-                .subscribeOn(schedulers.computeThreadScheduler)
-                .observeOn(schedulers.calleeThreadScheduler)
-                .subscribe()
-    )
+            )
+        }
+    }
 
-    fun newJob(modelName: String, version: String): SyftJob {
-        val job = SyftJob(modelName, version)
-        signallingClient.send(
-            REQUESTS.CYCLE_REQUEST,
-            CommunicationDataFactory.requestCycle(workerId, job, "", "", "")
-        )
-        workerJobs.add(job)
-        return job
+    @Synchronized
+    private fun setSyftWorkerId(workerId: String) {
+        if (!this::workerId.isInitialized)
+            this.workerId = workerId
     }
 
+    private fun getPing() = ""
+    private fun getDownloadSpeed() = ""
+    private fun getUploadSpeed() = ""
+
     private fun handleResponse(response: SocketResponse) {
         when (response.data) {
-            is AuthenticationSuccess ->
+            is AuthenticationSuccess -> {
                 this.workerId = response.data.workerId
-            is CycleResponseData -> {
-                when (response.data) {
-                    is CycleResponseData.CycleAccept -> "accept here"
-                    is CycleResponseData.CycleReject -> "set timeout for job"
-                }
+
             }
+            is CycleResponseData -> handleCycleResponse(response.data)
+            is ReportResponse -> Log.i(TAG, response.data.status)
+        }
+    }
 
+    private fun handleCycleResponse(responseData: CycleResponseData) {
+        when (responseData) {
+            is CycleResponseData.CycleAccept -> {
```

Comment:
I would have put the logic for each responseData in a separate function for readability. Minor cosmetics